substrate/runtime/contract/src/vm/prepare.rs

```diff
@@ -0,0 +1,285 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate. If not, see <http://www.gnu.org/licenses/>.
+
+//! Module that takes care of loading, checking and preprocessing of a
+//! wasm module before execution.
```

Comment:
this code mostly preserved from [`src/vm.rs`](https://github.com/paritytech/substrate/blob/690f52f5ab269b6c35aba9e7056f3362291c7f5a/substrate/runtime/contract/src/vm.rs)

the difference in that instead of scanning imports [for imports of memory instances](https://github.com/paritytech/substrate/blob/690f52f5ab269b6c35aba9e7056f3362291c7f5a/substrate/runtime/contract/src/vm.rs#L555) we now also process all functions, [checking their signatures](https://github.com/paritytech/substrate/pull/478/files#diff-0b88a741560f3ad7f23f1080a359d02fR105).